### DISCUSSION TITLE
1.) API Data are set twice on mount - 2.) Favorite Button not working on updated API Data / API Data don't get passed down to other routes when updated

### DISCUSSION THREAD

**Discussion by thanhtran-git** [neuefische/questions#813] [Web] [answered]
Good morning,

You don't have to answer both questions at once. **_The second one is more important ☺️_**

In our project, we are using data from an API to display Event Cards on the home page, in the favorites list and as search results. 

# **Problem 1: API Data are set twice on mount** 
For the purpose of this question, we will display 10 Event Cards at initial page load on the homepage at "/". 

Console.log("API Data from app.js: ", apiData) should display an array of 10 objects, but instead, it's an array 20 objects. 
The 10 results just got set twice and put into the apiData state. 

Note: I'm using `setApiData((prev) => [...prev, ...apiData.......... ])`  to extend the list of results with every click on a "See More" button, which loads a new batch of 10 event cards. Because of the double setting of the state on the first page load, the first 10 results get loaded twice into the state. 

I know that react strictmode renders components twice, but here it's causing confusion.
In the _app.js, I've added  `  const [counter, setCounter] = useState(0);` to check whether the page load  causes the counter to increment twice.

`  useEffect(() => {
    setCounter((prev) => prev + 1);
  }, []);`
`      <p>rendered {counter} times on mount</p>`

Expected behaviour: the counter should display 1 because the useEffect and setCounter run once  on mount .
![image](https://github.com/neuefische/questions/assets/85113171/2eda2367-ee3d-47d7-b8e5-03661b08ddca)

But, the counter displays  "2" which means the component rendered twice. And that means, the initially fetched API data got set twice.

## Question
What causes that? How to prevent it without having to set strictmode to false? 

# **2. Problem:  Favorite Button doesn't work on updated API Data / API Data don't get passed down as props to other routes**
**Note:** I have disabled strictmode for this question else the console.logs would be an absolute mess to deal with.

Every Event Card contains a Star Button to mark the Event as Favorite. Doing so should save the Card in the /favorites route.

Marking any of the first 10 results as favorite properly displays them as Favorite in that route:

But when we scroll down to the the See More Button and click it to load the next 10 results, marking any of those will not display them as favorite in the /favorites route:

The way the Favorites page works is, it filters through the API Data it received as props and checks whether the IDs of the favorited Cards match the ones of the apiData or not. 

When clicking on the "See More button", the state variable apiData is updated and increased by 10 more results, now at 20 results.

But when I check the apiData of the /favorites route with `  console.log("API Data from FAVORITES", apiData);`, the results are still at 10. 
It doesn't matter how many more times I load more data. The apiData state in the _app.js can even contain 50 or more results, while  in /favorites the console.log only shows the initial 10 results. 

That means the apiData is not passed down to the /favorites route properly. And that makes it impossible  for the route to filter through the apiData to display newly marked Favorites. Addtionally, any other state like searchData, the results of the search, will also not get passed downs as prop as well, so that search results also cannot be favorited.

![image](https://github.com/neuefische/questions/assets/85113171/c5515ad5-785c-4845-ab1e-b158752a2fd3)

I would avoid useSWR as much as possible, because it would make the code of my dynamic API fetch more complicated than it already is, and I only have this week left to finish the project.

@onemanwenttomow You have helped with this buggy behavior before but it occured again. If you have spare time left I would appreciate a ton.
@moonwave99 @doemser @JessicaLoers or any other coach 👋🏻 

## Links
- [PR](https://github.com/jessicareinh/capstone-event-app/pull/43)
- [Codespace](https://studious-spork-rvgp66w56jrfprv.github.dev/)
- [Preview](https://studious-spork-rvgp66w56jrfprv-3000.app.github.dev/)

**Comment by thanhtran-git** [comment, marked as answer]
TLDR answer of question 2: 
![mario-anchor-meme](https://github.com/neuefische/questions/assets/85113171/6da5b29c-aac7-469b-85ac-eda4bb82cf8f)

For anyone interested:
Instead of Nextjs links, I was using regular HTML anchor tags (styled.a) in the navbar.
Using those <a> within Next.js will generally trigger a full page refresh by default when clicked, potentially causing the API to be fetched again. This is Standard behavior: Anchor tags by their nature are designed to navigate to a new URL, which in most cases involves a full page load to fetch the requested resource.
In my case, the favorite page always did a full refresh refresh including a fetch of the initial 10 event cards. It never got the chance to receive the newly loaded apiData.

In contrast, using Next Link will not trigger a full page reload. They allow smooth navigation through the page without losing the application state.

Big Thanks @ahohnsen & @JessicaLoers
